presto-main/src/main/java/io/prestosql/execution/QueryTracker.java

```diff
@@ -176,15 +177,20 @@ private void enforceTimeLimits()
             if (query.isDone()) {
                 continue;
             }
-            Duration queryMaxRunTime = getQueryMaxRunTime(query.getSession());
-            Duration queryMaxExecutionTime = getQueryMaxExecutionTime(query.getSession());
-            Optional<DateTime> executionStartTime = query.getExecutionStartTime();
-            DateTime createTime = query.getCreateTime();
-            if (executionStartTime.isPresent() && executionStartTime.get().plus(queryMaxExecutionTime.toMillis()).isBeforeNow()) {
-                query.fail(new PrestoException(EXCEEDED_TIME_LIMIT, "Query exceeded the maximum execution time limit of " + queryMaxExecutionTime));
+            try {
+                Duration queryMaxRunTime = getQueryMaxRunTime(query.getSession());
+                Duration queryMaxExecutionTime = getQueryMaxExecutionTime(query.getSession());
+                Optional<DateTime> executionStartTime = query.getExecutionStartTime();
+                DateTime createTime = query.getCreateTime();
+                if (executionStartTime.isPresent() && executionStartTime.get().plus(queryMaxExecutionTime.toMillis()).isBeforeNow()) {
+                    query.fail(new PrestoException(EXCEEDED_TIME_LIMIT, "Query exceeded the maximum execution time limit of " + queryMaxExecutionTime));
+                }
+                if (createTime.plus(queryMaxRunTime.toMillis()).isBeforeNow()) {
+                    query.fail(new PrestoException(EXCEEDED_TIME_LIMIT, "Query exceeded maximum time limit of " + queryMaxRunTime));
+                }
             }
-            if (createTime.plus(queryMaxRunTime.toMillis()).isBeforeNow()) {
-                query.fail(new PrestoException(EXCEEDED_TIME_LIMIT, "Query exceeded maximum time limit of " + queryMaxRunTime));
+            catch (Exception e) {
```

Comment:
What are we guarding against here?

Reply:
This PR is intended to be an extension of change https://github.com/prestosql/presto/pull/3621/files (which fixed stats calculation error for toralBufferedBytes). It ensures if any other stat calculation fails (maybe because of negative value exception etc), then it will ensure to fail that query.

Reply:
Do we know under what circumstances such calculation could fail?

Reply:
No, till now, we don't really know the root cause. But, we have seen the issue of negative value exception while calculating toralBufferedBytes sometimes.

Reply:
Do we have an issue for that?

Reply:
We got the issue for matric totalBufferBytes. I could find the issue: https://github.com/prestosql/presto/issues/696 which failed with the same exception but for a different metric. So, this change is just a safeguard for any remaining metric for which we didn't fix.

Reply:
I do not want Presto users to become used to Presto throwing warnings and let them think "oh, this is a warning, i can ignore it".

I am concerned that if we add a safeguard, we won't learn about those potential failures and we won't be able to fix them.

Reply:
Your concern is valid. But, this fix will at least make sure that the problematic query fails and not get queued forever. If there is something wrong with stats calculation, it doesn't even show it on UI as UI too is based on building query Info.

Reply:
I don't see how you would get an exception in this code.  Do you have a stack trace that shows a problem in this code?

Reply:
I got this issue when I was running prestodb. Here is the stack trace:

```
ERROR   ResourceGroupManager    com.facebook.presto.execution.resourceGroups.InternalResourceGroupManager       Exception while processing queued queries for RootInternalResourceGroup{id=global}
java.lang.IllegalArgumentException: size is negative
        at io.airlift.units.Preconditions.checkArgument(Preconditions.java:26)
        at io.airlift.units.DataSize.<init>(DataSize.java:58)
        at io.airlift.units.DataSize.succinctDataSize(DataSize.java:48)
        at io.airlift.units.DataSize.succinctBytes(DataSize.java:43)
        at com.facebook.presto.execution.StageStateMachine.getStageInfo(StageStateMachine.java:347)
        at com.facebook.presto.execution.SqlStageExecution.getStageInfo(SqlStageExecution.java:276)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at com.google.common.collect.CollectSpliterators$1WithCharacteristics.lambda$forEachRemaining$1(CollectSpliterators.java:64)
        at java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:110)
        at com.google.common.collect.CollectSpliterators$1WithCharacteristics.forEachRemaining(CollectSpliterators.java:64)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
        at com.facebook.presto.execution.scheduler.SqlQueryScheduler.getStageInfo(SqlQueryScheduler.java:405)
        at com.facebook.presto.execution.SqlQueryExecution.buildQueryInfo(SqlQueryExecution.java:875)
        at com.facebook.presto.execution.SqlQueryExecution.getQueryInfo(SqlQueryExecution.java:839)
        at com.facebook.presto.execution.resourceGroups.InternalResourceGroup.enforceTimeLimits(InternalResourceGroup.java:972)
        at com.facebook.presto.execution.resourceGroups.InternalResourceGroup$RootInternalResourceGroup.processQueuedQueries(InternalResourceGroup.java:1167)
        at com.facebook.presto.execution.resourceGroups.InternalResourceGroupManager.refreshAndStartQueries(InternalResourceGroupManager.java:228)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

Reply:
is it conceivable we've fixed that over past two years?

Reply:
It was fixed by https://github.com/prestosql/presto/pull/3621/files, but I thought the same issue might arise for a different metric, so I raised PR.

Reply:
I don't think this can happen here, so I'd rather not have the extra checks.